### PR TITLE
feat: add worklist

### PR DESCRIPTION
**Pull request by simbig**
- [x] Added automated tests
- [x] Documented for all relevant versions
- [x] Updated the changelog

**Changes**

- Add classes and methods for building pipetting instructions for Tecan Freedom EVO robots in the Gemini WorkList file format (*.gwl).

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -16,7 +16,8 @@
   },
   "require": {
     "php": "^7.4 || ^8",
-    "illuminate/support": "^8 || ^9",
+    "bensampo/laravel-enum": "^5.3",
```
Comment (spawnia) [inline composer.json]:
```suggestion
    "bensampo/laravel-enum": "^2 || ^3 || ^4 || ^5",
```

Necessary to match illuminate/support

File: phpstan.neon [hunk 2]
```diff
@@ -4,3 +4,4 @@ parameters:
 	- src
 	- tests
 	tmpDir: .build/phpstan/
+	reportMaybesInPropertyPhpDocTypes: false
```
Comment (spawnia) [inline phpstan.neon]: Why?

File: src/Tecan/BasicCommands/BasicPipettingActionCommand.php [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+    protected function getTipMask(): string
+    {
+        return $this->tipMask;
+    }
+
+    public function setTipMask(int $tipMask): void
+    {
+        $this->tipMask = (string) $tipMask;
+    }
```
Comment (spawnia) [inline src/Tecan/BasicCommands/BasicPipettingActionCommand.php:44]: Why not just make this property a `public int`?

File: src/Tecan/BasicCommands/BreakCommand.php [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+class BreakCommand implements Command
+{
+    public static function commandLetter(): string
+    {
+        return 'B;';
+    }
+
+    public function toString(): string
+    {
+        return $this::commandLetter();
```
Comment (spawnia) [inline src/Tecan/BasicCommands/BreakCommand.php]:
```suggestion
        return static::commandLetter();
```

Comment (spawnia) [reply src/Tecan/BasicCommands/BreakCommand.php]: The same applies in more places

File: src/Tecan/Location/Location.php [hunk 5]
```diff
@@ -0,0 +1,16 @@
+<?php declare(strict_types=1);
+
+namespace Mll\LiquidHandlingRobotics\Tecan\Location;
+
+interface Location
```
Comment (spawnia) [inline src/Tecan/Location/Location.php:5]: Implementors often have to return empty string as default values for some methods, that indicates the interface is not well-factored

File: src/Tecan/TecanException.php [hunk 6]
```diff
@@ -0,0 +1,7 @@
+<?php declare(strict_types=1);
+
+namespace Mll\LiquidHandlingRobotics\Tecan;
+
+final class TecanException extends \Exception
```
Comment (spawnia) [inline src/Tecan/TecanException.php:5]: We should decide if we want to use `final` and do it throughout

File: src/Tecan/TecanProtocol.php [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    /**
+     * @var Collection<Command>
+     */
+    private Collection $commands;
+
+    private TipMask $tipMask;
+
+    public function __construct(
+        TipMask $tipMask
+    ) {
```
Comment (spawnia) [inline src/Tecan/TecanProtocol.php]:
```suggestion
    public function __construct(TipMask $tipMask) {
```

File: src/Tecan/TecanProtocol.php [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    {
+        return $this
+            ->commands
+            ->map(
+                function ($command): string {
+                    assert($command instanceof Command);
+
+                    return $command->toString();
+                }
+            )->join(self::NEW_LINE) . self::NEW_LINE;
```
Comment (spawnia) [inline src/Tecan/TecanProtocol.php]:
```suggestion
            ->map(function (Command $command): string {
                return $command->toString();
            })
            ->join(self::NEW_LINE)
            . self::NEW_LINE;
```

File: src/Tecan/TecanProtocol.php [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+use Illuminate\Support\Collection;
+use Mll\LiquidHandlingRobotics\Tecan\BasicCommands\BreakCommand;
+use Mll\LiquidHandlingRobotics\Tecan\BasicCommands\Command;
+use Mll\LiquidHandlingRobotics\Tecan\BasicCommands\PipettingActionCommand;
+use Mll\LiquidHandlingRobotics\Tecan\TipMask\TipMask;
+
+class TecanProtocol
+{
+    public const NEW_LINE = "\r\n";
```
Comment (spawnia) [inline src/Tecan/TecanProtocol.php]:
```suggestion
    /**
     * Tecan software runs on Windows.
     */
    public const WINDOWS_NEW_LINE = "\r\n";
```

File: src/Tecan/TipMask/TipMask.php [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+    public ?int $currentTip = null;
+
+    abstract public function isLastTip(): bool;
+
+    public function nextTip(): int
+    {
+        if (! isset($this->currentTip) || $this->isLastTip()) {
+            $this->currentTip = self::FIRST_TIP;
+        } else {
+            // because of the bitwise nature we can just multiply current tip with 2 switch to the next tip
```
Comment (spawnia) [inline src/Tecan/TipMask/TipMask.php]: Grammar seems off

File: tests/Unit/Tecan/BasicCommands/AspirateTest.php [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+namespace Mll\LiquidHandlingRobotics\Tests\Unit\Tecan\BasicCommands;
+
+use Mll\LiquidHandlingRobotics\Tecan\BasicCommands\Aspirate;
+use Mll\LiquidHandlingRobotics\Tecan\Location\BarcodeLocation;
+use Mll\LiquidHandlingRobotics\Tecan\Location\PositionLocation;
+use Mll\LiquidHandlingRobotics\Tests\TestImplentations\TestLiquidClass;
+use Mll\LiquidHandlingRobotics\Tests\TestImplentations\TestRack;
+use PHPUnit\Framework\TestCase;
+
+class AspirateTest extends TestCase
```
Comment (spawnia) [inline tests/Unit/Tecan/BasicCommands/AspirateTest.php]: Let's definitely use `final` for all test classes

File: tests/Unit/Tecan/TecanProtocolTest.php [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+    public function testAspirateWithBarcodeLocationForTips(): void
+    {
+        $tecanProtocol = new TecanProtocol(new TipMaskFourTips());
+
+        $liquidClass = new TestLiquidClass();
+        $rack = new TestRack();
+        $barcodeLocation = new BarcodeLocation('barcode', $rack);
+        $location = new BarcodeLocation('barcode1', $rack);
+
+        for ($i = 1; $i <= 5; ++$i) {
```
Comment (spawnia) [inline tests/Unit/Tecan/TecanProtocolTest.php]:
```suggestion
        foreach (range(1, 5) as $_) {
```

File: src/Tecan/TecanProtocol.php (hunk 8 above)
Comment (simbig) [reply src/Tecan/TecanProtocol.php]: this is my workaround for this error:
image
Any idea how to solve it?

File: src/Tecan/BasicCommands/BasicPipettingActionCommand.php (hunk 3 above)
Comment (simbig) [reply src/Tecan/BasicCommands/BasicPipettingActionCommand.php:44]: those funtions are required by and interface. Properties can not be required by an interface

File: src/Tecan/Location/Location.php (hunk 5 above)
Comment (simbig) [reply src/Tecan/Location/Location.php:5]: Its due to the nature of the file used by the robot. It requires the "position" to be an empty string if an barcode is used an vice-versa. Thus the empty string is considers as an empty/not-set value here.
I could extract that to an constant if it is helpful for understanding

Comment (spawnia) [reply src/Tecan/Location/Location.php:5]: I would prefer `null` while we are in PHP, only serializing it to an empty string at the edge of the system.

File: src/Tecan/TecanProtocol.php (hunk 8 above)
Comment (spawnia) [reply src/Tecan/TecanProtocol.php]: `Collection` from Laravel 9 expects 2 generic arguments.

**Review by spawnia — APPROVED** [review]

**Pull request merged**